Sources/PackageCollections/PackageIndexAndCollections.swift

```diff
@@ -0,0 +1,284 @@
+                    )
+                ))
+            case .failure(let indexError):
+                switch collectionsResult {
+                case .success(let metadataResult):
+                    // Metadata from `PackageCollections`, which is a combination of
+                    // package index/alternative (e.g., GitHub) and collection data.
+                    callback(.success(
+                        PackageCollectionsModel.PackageMetadata(
+                            package: metadataResult.package,
+                            collections: metadataResult.collections,
+                            provider: metadataResult.provider
+                        )
+                    ))
+                case .failure(let collectionsError):
+                    // Failed to get metadata through `PackageIndex` and `PackageCollections`.
+                    // Return index's error unless no index is configured.
+                    switch indexError {
+                    case PackageIndexError.featureDisabled, PackageIndexError.notConfigured:
+                        callback(.failure(collectionsError))
+                    default:
+                        callback(.failure(indexError))
+                    }
+                }
+            }
+        }
+    }
+    
+    public func findPackages(
+        _ query: String,
+        collections: Set<PackageCollectionsModel.CollectionIdentifier>?,
+        callback: @escaping (Result<PackageCollectionsModel.PackageSearchResult, Error>) -> Void
+    ) {
+        self.index.findPackages(query) { indexResult in
+            switch indexResult {
+            case .failure(PackageIndexError.featureDisabled), .failure(PackageIndexError.notConfigured):
+                self.collections.findPackages(query, collections: collections, callback: callback)
+            case .failure(let error):
+                callback(.failure(error))
+            case .success(let indexSearchResult):
+                // For each package in the search result, find the collections that it belongs to.
+                self.collections.listPackages(collections: collections) { collectionsResult in
+                    switch collectionsResult {
+                    case .failure:
+                        callback(.success(indexSearchResult))
+                    case .success(let collectionsSearchResult):
+                        let items = indexSearchResult.items.map { item in
+                            PackageCollectionsModel.PackageSearchResult.Item(
+                                package: item.package,
+                                collections: collectionsSearchResult.items.first(where: {
+                                    item.package.identity == $0.package.identity && item.package.location == $0.package.location
+                                })?.collections ?? [],
+                                indexes: item.indexes
+                            )
+                        }
```

Comment:
does this "decorate" the index search results with information about which collection the result belongs to?